libbpf-tools/tcpdrop.bpf.c

```diff
@@ -0,0 +1,180 @@
+#include <vmlinux.h>
+
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_endian.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+
+#include "tcpdrop.h"
+
+#ifndef AF_INET
+#define AF_INET 2
+#endif
+#ifndef AF_INET6
+#define AF_INET6 10
+#endif
+#ifndef ETH_P_IP
+#define ETH_P_IP 0x0800
+#endif
+#ifndef ETH_P_IPV6
+#define ETH_P_IPV6 0x86dd
+#endif
+
+struct {
+	__uint(type, BPF_MAP_TYPE_RINGBUF);
+	__uint(max_entries, 512);
+} events SEC(".maps");
+
+#define MAX_STACK_DEPTH 15
+struct {
+	__uint(type, BPF_MAP_TYPE_STACK_TRACE);
+	__uint(max_entries, 512);
+	__uint(key_size, sizeof(u32));
+	__uint(value_size, MAX_STACK_DEPTH * sizeof(u64));
+} stack_traces SEC(".maps");
+
+const volatile char ipv4_only = 0;
+const volatile char ipv6_only = 0;
+const volatile __u32 netns_id = 0;
+
+SEC("tracepoint/skb/kfree_skb")
+int tp__skb_free_skb(struct trace_event_raw_kfree_skb *args)
+{
+	struct event *event;
+	struct sk_buff *skb;
+	struct sock *sk;
+	void *head;
+	u16 network_header, transport_header, protocol;
+	u64 pid_tgid;
+	u32 pid;
+	struct net *net;
+	u32 inum;
+	u8 state;
+	struct iphdr ip;
+	struct ipv6hdr ip6;
+	struct tcphdr tcp;
+
+	skb = args->skbaddr;
+	if (!skb) {
+		return 0;
+	}
+
+	event = bpf_ringbuf_reserve(&events, sizeof(*event), 0);
+	if (!event) {
+		return 0;
+	}
+
+	if (bpf_core_field_exists(args->reason)) {
+		if (args->reason <= SKB_DROP_REASON_NOT_SPECIFIED) {
+			bpf_ringbuf_discard(event, 0);
+			return 0;
+		}
+		event->drop_reason = args->reason;
+	} else {
+		event->drop_reason = -1;
+	}
+
+	if (bpf_ringbuf_query(&events, BPF_RB_AVAIL_DATA) >= 511) {
+		bpf_ringbuf_discard(event, 0);
+		return 0;
+	}
+
+	sk = NULL;
+	bpf_core_read(&sk, sizeof(sk), &skb->sk);
+
+	if (bpf_core_read(&head, sizeof(head), &skb->head) ||
+	    bpf_core_read(&network_header, sizeof(network_header),
+			  &skb->network_header) ||
+	    bpf_core_read(&transport_header, sizeof(transport_header),
+			  &skb->transport_header)) {
+		bpf_ringbuf_discard(event, 0);
+		return 0;
+	}
+
+	protocol = args->protocol;
+	if (protocol != ETH_P_IP && protocol != ETH_P_IPV6) {
+		bpf_ringbuf_discard(event, 0);
+		return 0;
+	}
+	if (ipv4_only && protocol != ETH_P_IP) {
+		bpf_ringbuf_discard(event, 0);
+		return 0;
+	}
+	if (ipv6_only && protocol != ETH_P_IPV6) {
+		bpf_ringbuf_discard(event, 0);
+		return 0;
+	}
```

Comment:
Check these before `bpf_ringbuf_reserve()` so that we don't have to use many bpf_ringbuf_discard() in each branch.